plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,425 @@
+
+	if (!mediatek_scaler_device_ddc_read(self, buf, sizeof(buf), recv, &recvsz, error))
+		return FALSE;
+
+	if (recvsz >= 4) {
+		g_autofree guint8 *ver = g_new0(guint8, 4);
+		g_autofree guint32 *ver32 = (guint32 *)ver;
+		memcpy(ver + 0, &recv[4], sizeof(guint8));
+		memcpy(ver + 1, &recv[5], sizeof(guint8));
+		memcpy(ver + 2, &recv[2], sizeof(guint8));
+		memcpy(ver + 3, &recv[3], sizeof(guint8));
+
+		fu_device_set_version_from_uint32(FU_DEVICE(self), *ver32);
+		return TRUE;
+	}
+	return FALSE;
+}
+
+static gboolean
+mediatek_scaler_device_probe_hardware_version(FuDevice *device, GError **error)
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	g_autofree guint8 *recv = g_malloc0(10);
+	guint8 buf[] = {0xCC, 0xFF, 0x00};
+	guint recvsz = 0;
+
+	if (!mediatek_scaler_device_ddc_read(self, buf, sizeof(buf), recv, &recvsz, error))
+		return FALSE;
+
+	if (recvsz >= 4) {
+		g_autofree guint8 *ver = g_new0(guint8, 4);
+		g_autofree guint32 *ver32 = (guint32 *)ver;
+		memcpy(ver + 0, &recv[4], sizeof(guint8));
+		memcpy(ver + 1, &recv[5], sizeof(guint8));
+		memcpy(ver + 2, &recv[2], sizeof(guint8));
+		memcpy(ver + 3, &recv[3], sizeof(guint8));
+
+		self->hw_ver = *ver32;
+		return TRUE;
+	}
+	return FALSE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_probe(FuDevice *device, GError **error)
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	g_autoptr(FuUdevDevice) udev_parent = NULL;
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autofree gchar *vendor_id = NULL;
+	g_autofree gchar *vid = NULL;
+	g_autofree gchar *pid = NULL;
+	g_autofree gchar *hw_ver = NULL;
+
+	/* FuUdevDevice->probe */
+	if (!FU_DEVICE_CLASS(fu_mediatek_scaler_device_parent_class)->probe(device, error))
+		return FALSE;
+
+	/* determine the i2c_dev for dp aux dev */
+	if (!fu_mediatek_scaler_device_use_aux_dev(device, error))
+		return FALSE;
+
+	/* open device */
+	if ((locker = fu_device_locker_new(self->i2c_dev, error)) == NULL)
+		return FALSE;
+
+	/* set the slave address */
+	if (!mediatek_scaler_ensure_device_address(self, DDC_DEST_ADDR >> 1, error))
+		return FALSE;
+
+	/* mediatek display is connected */
+	if (!mediatek_scaler_display_is_connected(self, error))
+		return FALSE;
+
+	/* set hardware version */
+	if (!mediatek_scaler_device_probe_hardware_version(device, error))
+		return FALSE;
+
+	/* set vid and pid from PCI bus */
+	udev_parent = fu_udev_device_get_parent_with_subsystem(FU_UDEV_DEVICE(device), "pci");
+	if (udev_parent == NULL)
+		return FALSE;
+
+	if (!fu_device_probe(FU_DEVICE(udev_parent), error))
+		return FALSE;
+
+	vid = g_strdup_printf("%04X", fu_udev_device_get_subsystem_vendor(udev_parent));
+	pid = g_strdup_printf("%04X", fu_udev_device_get_subsystem_model(udev_parent));
+
+	/* add IDs */
+	vendor_id = g_strdup_printf("PCI:0x%s", vid);
+	fu_device_add_vendor_id(device, vendor_id);
+
+	fu_device_add_instance_str(device, "VID", vid);
+	fu_device_add_instance_str(device, "PID", pid);
```

Comment:
i wonder if fu_device_incorporate might make your life a lot easier here

Reply:
You mean to incorporate the parent udevice?

Reply:
the udev_parent, yep